### DISCUSSION TITLE
Define a place where to setup a `DataUpdateCoordinator`

### DISCUSSION THREAD

**Discussion by zweckj** [home-assistant/architecture#1073] [General] [answered]
Currently, when integrations need to setup their `DataUpdateCoordinators`, they don't have a clearly defined place to execute logic that should only be executed once during initial setup. 

On a 5 minute scan I found:

- teslemetry
- renault
- ridwell

which are directly overriding `async_config_entry_first_refresh` and thus not benefitting from the usual exception handling.

- lamarzocco
- ring
- systemmonitor
- opower

which are using some kind of `first_refresh` variable

Then there are integrations like (might not be complete)
- switchbee
- tedee

which are loading things only once in the `async_update` by determining if a certain variable is `None`

- anova
- enphase_envoy
- livisi
- shelly
- synology_dsm
- tankerkoenig
- template
- webmin

wrote their own `async_setup` method.

Additionaly, some integrations will do that initial setup in `async_entry_setup` and pass then the initalized things to the `coordinator` (nothing wrong with that, but will need to unnecessary parameters), but this is not possible for every integration, if they need access to things (variables/methods) from the `coordinator`.

My proposal would be to add an empty method `_async_setup` (open for suggestions here) to `async_config_entry_first_refresh`, which developers can override in their implementations to have a place which is properly exception handled to do coordinator setup/load elements which only need to be loaded once.

Example implementation PR: https://github.com/home-assistant/core/pull/114856

**Comment by balloob** [comment]
I think that's a good idea. It will also encourage people to use `async_config_entry_first_refresh`.

**Comment by gjohansson-ST** [comment]
This would likely be very useful and can replace most integrations specific implementations I would guess

In difference to the "normal" `async_refresh" this method (in your example PR) isn't returning data but only a bool.
Should this initial setup not also be able to set data, it needs to be handled manually, or handle it like `async_refresh`?

**Reply by zweckj** [reply]
Since it is run in `async_config_entry_first_refresh`, it doesn’t need to return data, since this will happen from the update directly afterwards anyways (in the „good“ case). And the method which returns the `bool` is just the wrapper for the actual update method, which returns `None`

**Reply by gjohansson-ST** [reply]
My thinking was that this could be used to initialize `self.data` with something which only needs to be updated once (such as `last_boot` in `systemmonitor`).
With your reasoning that would have to remain in `_async_update_data` and as such we would have to keep the initial variable and compute if it's first refresh or not and as such this implementation has no benefit for that particular integration.

I would argue that we could embed so `self.data` is also populated from this "run once" method and make it more useful.
Could also be done "manually" within the method obviously so still returning a `bool`

Only consideration maybe that `_async_update_data` might overwrite it so devs would have to consider this in it's implementation if that's the case.

**Reply by zweckj** [reply]
I don’t get it. You could just set `self.last_boot` in the `_async_update` without worrying that it’s overwritten again. I think the `self.data` is normally different data than the one that is only updated once. 
Again, the `bool` method is only the wrapper for exception handling, not the one that should be overwritten/implemented.

**Reply by gjohansson-ST** [reply]
[This line](https://github.com/home-assistant/core/pull/114856/files#diff-c274e5f7f48b4f740b17d6053cfc5019c41576ef1323770ae24ddfb3759561adR293) could say `self.data = await self._async_setup()` to load initial data (same principle which is used in `async_refresh`).

It would just make it more obvious if `async_setup` isn't just returning `None` but can return what `self.data` expects and load data which you only update once (my example using `systemmonitor` previously).

Anyhow I think it would be good to have a "run once" option during setup so I like this proposal

**Reply by zweckj** [reply]
Ok coming back to this. My problem with returning `self.data` is, that it's less flexible imo. If I return `None` you can just set `self.data = ...` during `_async_setup`. If I force it to return the data type you always have to return something, limiting the use cases where you don't need to.

**Comment by joostlek** [comment]
I think it makes sense to have a function for the setup. Sometimes you just need to have a place to request something only once, for example requesting some kind of user or device data, that doesn't update as often as the DUC polls. I like this 👍🏻

**Comment by frenck** [comment, marked as answer]
This proposal has been discussed and approved.

✅ 

../Frenck
